### ISSUE TITLE
Why YOLO11-n performs much worse in Ultralytics 8.4 than 8.3？

### ISSUE THREAD

**Issue by guan2000910** [ultralytics/ultralytics#24638]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I train YOLO11-n on a simple single-class dataset of 10k images, one object per image. The model performs poorly in Ultralytics 8.4, much worse than in 8.3. What causes this?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @guan2000910, thank you for reporting this and for your interest in Ultralytics 🚀! This is an automated response to help gather the details needed for troubleshooting, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this appears to be a performance regression question between package versions, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) 🐛 so the team can compare `8.3` vs `8.4` under the same conditions.
If this is a custom training ❓ question, please provide as much information as possible, including dataset image examples, training logs, exact commands used, model/config files, and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by venu-banaras** [comment]
Please share the training parameters and evaluation metrics with us too.

**Comment by raimbekovm** [comment]
to look into this we need concrete numbers and a way to reproduce. please share:

1. **exact versions** for both runs: output of `pip show ultralytics torch` from each environment
2. **the training commands or scripts** (identical except for the package version), including any non-default hyperparameters and the seed if set
3. **a side-by-side metric table** from the printed val output of each run, e.g.:

```
ultralytics 8.3.X: mAP50=..., mAP50-95=..., final box_loss=...
ultralytics 8.4.Y: mAP50=..., mAP50-95=..., final box_loss=...
```

4. **the dataset yaml** (paths can be redacted, structure and class definition matter)
5. **training logs from epoch 1 and the final epoch** of each run (a few lines each is enough)

if possible, a single-image overfit test (10 epochs on one image) reproducing the gap would be the fastest way to isolate the regression.

**Comment by guan2000910** [comment]
@venu-banaras ,@raimbekovm  Sorry, I conduct experiments in an intranet environment, so I cannot share any relevant data. I am doing segmentation experiments with YOLO11-n. Both experiments adopt identical training configurations, and both achieve around 0.99 mAP after training. However, there is a huge difference in prediction results. There is only one class in dataset annotations: ground，the masks predicted in Ultralytics 8.4 always lack the lower part of objects in the images.

**Comment by raimbekovm** [comment]
the data itself is not strictly required. what would unblock this:

1. exact `pip show ultralytics` from both runs (e.g. `8.3.253` vs `8.4.X`)
2. typical input image **dimensions and aspect ratio** (W×H), since letterbox padding behaves differently for portrait/landscape vs square images
3. whether you call `model.predict` directly or wrap it, and whether you set any non-default predict args
4. a **side-by-side prediction screenshot** on one image (your own or a public one) showing the same mask cut at the bottom in 8.4 but full in 8.3. one redacted PNG of the mask overlay is enough, the image content does not matter

as a quick local test, run prediction once with `retina_masks=True` and once with the default in 8.4. if the truncation disappears with `retina_masks=True`, the regression is in the non-native mask path. if it stays, it is elsewhere. that single experiment narrows it down significantly.

**Comment by guan2000910** [comment]
@raimbekovm <img width="1920" height="1440" alt="Image" src="https://github.com/user-attachments/assets/2a62de31-d3ff-478b-8b23-3a53c0ed9d7a" />

<img width="1920" height="1440" alt="Image" src="https://github.com/user-attachments/assets/62d202b7-0e9f-4581-980f-1ee835fcdcd9" />

he mask comparison results are shown above.
Versions used: 8.3.221 and 8.4.56.
Original image resolution: 640×480.
I trained the model directly using model.train() with batch size set to 512 and input image size imgsz=640, keeping all other training parameters as default.
During predict, I extracted and drew masks directly from results[0].masks without any extra custom processing.

**Comment by raimbekovm** [comment]
quick update: i ran the exact pipeline diff on `yolo11n-seg` + a 640×480 image in `v8.3.221` and `8.4.56`. the predict path (NMS, `process_mask`, `scale_masks`, `crop_mask`) produces **byte-identical** masks, boxes, and scores between the two versions. so the regression is not in the shared postprocess.

that leaves the training pipeline as the most likely source. a few segmentation-augmentation changes landed between 8.3 and 8.4 that touch how box edges are handled when masks cross image borders, which is exactly your `ground` case.

a small targeted test would isolate it for sure. could you cross-load weights between the two environments?

1. **load your 8.3-trained `.pt`** with `ultralytics 8.4.56` and predict
2. **load your 8.4-trained `.pt`** with `ultralytics 8.3.221` and predict

if (1) gives full-bottom masks and (2) gives truncated masks, the issue is in the trained weights (training pipeline). if it is the reverse, it is the predict path under a specific tensor layout. one screenshot from each is enough.

**Comment by guan2000910** [comment]
@raimbekovm Sorry, I cannot perform the second test. Models trained in 8.4 cannot be directly loaded and run in 8.3 environment. It will throw an error due to structural and version incompatibility. So cross-load validation in the 8.3 environment is not possible.When deploying the model trained with version 8.3 in Ultralytics 8.4 environment, the predicted masks are complete at the bottom.The YOLO11-n weights I used are not the latest ones. Could this be the cause?

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24655: Fix `segment2box` dropping polygon vertices on image border](https://github.com/ultralytics/ultralytics/pull/24655).

This change updates the boundary handling in `segment2box()` so segmentation points that lie exactly on the image edge are no longer discarded. Specifically, the inside-image checks in `ultralytics/utils/ops.py` were changed from strict boundary comparisons to inclusive ones, which should help avoid edge-aligned polygons producing slightly cropped or incorrect bounding boxes.

Credit goes to @raimbekovm for the fix, and thanks as well to everyone who helped report and investigate the issue.

If you'd like to test the change now, you can install the latest `main` branch with:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Or, if preferred, you can wait for the next official PyPI release.

Please let us know whether the merged changes resolve the issue on your side. Thanks 🙏 for reporting this, and if the problem still persists, any additional feedback or a minimal reproducible example would be very helpful.

**Comment by raimbekovm** [comment]
thanks for the report and for the follow-up details in the thread. the comparison helped us pin down the `segment2box()` regression, and the fix shipped in [#24655](https://github.com/ultralytics/ultralytics/pull/24655). do reopen if you still see truncation on your data once the next 8.4.x release lands.
